ckanext/datajson/tests/test_datajsonharvester.py

```diff
@@ -41,95 +37,470 @@ def setup(cls):
-        harvest_object, result, dataset, errors = self.run_source(url=url)
-
+        datasets = self.run_source(url=url)
+        dataset = datasets[0]
         # assert_equal(first element on list
         expected_title = "NCEP GFS: vertical profiles of met quantities at standard pressures, at Barrow"
         assert_equal(dataset.title, expected_title)
         tags = [tag.name for tag in dataset.get_tags()]
         assert_in(munge_title_to_name("ORNL"), tags)
         assert_equal(len(dataset.resources), 1)
-    
+
     def test_datason_usda(self):
         url = 'http://127.0.0.1:%s/usda' % mock_datajson_source.PORT
-        harvest_object, result, dataset, errors = self.run_source(url=url)
+        datasets = self.run_source(url=url)
+        dataset = datasets[0]
         expected_title = "Department of Agriculture Congressional Logs for Fiscal Year 2014"
         assert_equal(dataset.title, expected_title)
         tags = [tag.name for tag in dataset.get_tags()]
-        assert_in(munge_title_to_name("Congressional Logs"), tags)
         assert_equal(len(dataset.resources), 1)
+        assert_in(munge_title_to_name("Congressional Logs"), tags)
+
+    def test_datajson_collection(self):
+        """ harvest from a source with a parent in the second place
+            We expect the gather stage to re-order to the forst place """
+        url = 'http://127.0.0.1:%s/collection-1-parent-2-children.data.json' % mock_datajson_source.PORT
+        obj_ids = self.run_gather(url=url)
+
+        identifiers = []
+        for obj_id in obj_ids:
+            harvest_object = harvest_model.HarvestObject.get(obj_id)
+            content = json.loads(harvest_object.content)
+            identifiers.append(content['identifier'])
+
+        if p.toolkit.check_ckan_version(max_version='2.7.99'):
+            # at CKAN 2.3 with GSA ckanext-harvest fork we expect just parents
+            # after "parents_run" a new job will be raised for children
+            expected_obj_ids = ['OPM-ERround-0001']
+        else:
+            # We always expect the parent to be the first on the list
+            expected_obj_ids = ['OPM-ERround-0001', 'OPM-ERround-0001-AWOL', 'OPM-ERround-0001-Retire']
+        assert_equal(expected_obj_ids, identifiers)
+    
+    def test_harvesting_parent_child_collections(self):
+        """ Test that parent are beeing harvested first.
+            When we harvest a child the parent must exists
+            data.json from: https://www.opm.gov/data.json """
+
+        url = 'http://127.0.0.1:%s/collection-1-parent-2-children.data.json' % mock_datajson_source.PORT
+        obj_ids = self.run_gather(url=url)
+
+        if p.toolkit.check_ckan_version(max_version='2.7.99'):
+            # at CKAN 2.3 with GSA ckanext-harvest fork we expect just parents
+            # after "parents_run" a new job will be raised for children
+            assert_equal(len(obj_ids), 1)
+        else:
+            assert_equal(len(obj_ids), 3)
```

Comment:
oof, yeah, the version check makes the tests very hard to understand. Using multiple class implementations means you can write separate tests for each class. The tests become much more readable.

Since you've already written assertions for 2.3, it would be a nice to keep them :) but not necessary for this work :)